### PR TITLE
Use Patroni API to set bootstrap-only options.

### PR DESCRIPTION
**Pull request by alexeyklyukin**
Call Patroni API /config in order to set special options that are
ignored when set in the configuration file, such as max_connections.
Per https://github.com/zalando-incubator/postgres-operator/issues/297

### CODE REVIEW

File: pkg/cluster/sync.go [hunk 1]
```diff
@@ -229,6 +229,7 @@ func (c *Cluster) syncStatefulSet() error {
 	var (
 		podsRollingUpdateRequired bool
 	)
+	// the following code should not return before the end of this function in a non-error case
```
Comment (zerg-junior) [inline pkg/cluster/sync.go]: 1. "The following" == code until the end of the function?
2. Can the code leave the cluster in inconsistent state in the case of early non-error return ?

File: pkg/cluster/sync.go [hunk 2]
```diff
@@ -303,6 +312,42 @@ func (c *Cluster) syncStatefulSet() error {
...
+			podName := util.NameFromMeta(pod.ObjectMeta)
+			c.logger.Debugf("calling Patroni API on a pod %s to set the following Postgres options: %v",
+				podName, optionsToSet)
+			if err := c.patroni.SetPostgresParameters(&pod, optionsToSet); err == nil {
+				return nil
+			} else {
+				c.logger.Warningf("could not patch postgres parameters with a pod %s: %v", podName, err)
+			}
+		}
+		return fmt.Errorf("could not call Patroni API to set Postgres options: failed on %d pods", len(pods))
```
Comment (zerg-junior) [inline pkg/cluster/sync.go]: What exactly does this error message intend to say ? That if Patroni fails to apply a param, it fails on all pods at once ?

File: pkg/cluster/volumes.go [hunk 3]
```diff
@@ -103,10 +103,10 @@ func (c *Cluster) resizeVolumes(newVolume spec.Volume, resizers []volumes.Volume
 
 	for _, pv := range pvs {
 		volumeSize := quantityToGigabyte(pv.Spec.Capacity[v1.ResourceStorage])
-		if volumeSize > newSize {
-			return fmt.Errorf("cannot shrink persistent volume")
-		}
-		if volumeSize == newSize {
+		if volumeSize >= newSize {
+			if volumeSize > newSize {
+				c.logger.Warningf("cannot shrink persistent volume")
```
Comment (zerg-junior) [inline pkg/cluster/volumes.go]: Do we log somewhere in the neighborhood of this statement which exactly volumes/sizes are we talking about ?

File: pkg/util/patroni/patroni.go [hunk 4]
```diff
@@ -74,6 +69,28 @@ func (p *Patroni) Failover(master *v1.Pod, candidate string) error {
 
 		return fmt.Errorf("patroni returned '%s'", string(bodyBytes))
 	}
+	return nil
+}
+
+// Failover does manual failover via Patroni POST API call
```
Comment (zerg-junior) [inline pkg/util/patroni/patroni.go]: Manual as in "make Patroni do that " ?

File: pkg/util/patroni/patroni.go [hunk 5]
```diff
@@ -14,13 +14,15 @@ import (
...
 	failoverPath = "/failover"
+	configPath   = "/config"
 	apiPort      = 8008
 	timeout      = 30 * time.Second
 )
 
 // Interface describe patroni methods
 type Interface interface {
 	Failover(master *v1.Pod, candidate string) error
+	SetPostgresParameters(server *v1.Pod, options map[string]string) error
```
Comment (zerg-junior) [inline pkg/util/patroni/patroni.go]: Is this method dedicated only to params that require Postgres restart ?

File: pkg/cluster/sync.go (hunk 1 above)
Comment (alexeyklyukin) [reply pkg/cluster/sync.go]: The rolling update is applied at the end of the function, and a code that will return beforehand will miss the rolling update.

File: pkg/cluster/sync.go (hunk 2 above)
Comment (alexeyklyukin) [reply pkg/cluster/sync.go]: That calling Patroni API failed, and we tried %d pods.

File: pkg/cluster/volumes.go (hunk 3 above)
Comment (alexeyklyukin) [reply pkg/cluster/volumes.go]: Hm.. this change seems to be unrelated to the actual PR. Not sure how did it get there.

File: pkg/util/patroni/patroni.go (hunk 4 above)
Comment (alexeyklyukin) [reply pkg/util/patroni/patroni.go]: yes, 'manual' might be not the best name here.

File: pkg/util/patroni/patroni.go (hunk 5 above)
Comment (alexeyklyukin) [reply pkg/util/patroni/patroni.go]: Not necessary.

File: pkg/cluster/sync.go (hunk 2 above)
Comment (zerg-junior) [reply pkg/cluster/sync.go]: > we tried %d pods.

i first understood the original err msg as "update failed on all `len(pods)` pods"

**Comment by alexeyklyukin** [comment]
@zerg-junior I did fix some warning messages and comments per your review. Please, take another look.

**Comment by zerg-junior** [comment]
:+1:

**Comment by alexeyklyukin** [comment]
👍

**Pull request merged**